Load pipeline results

nicolafan opened the pull request:
### Link to the relevant feature request
<!--

Link to the issue describing the feature request. The affected feature request should have the label 'help wanted' or
'accepted'.

-->

#1192 

### Description of the Change

<!--

We must be able to understand the design of your change from this description, so please walk us through the concepts.

-->

Add the classmethod `load_from_directory` to the class `pykeen.pipeline.PipelineResult`. The class method produces an instance of `PipelineResult` containing the information stored in the directory produced via `save_to_directory`.

Not all the instance attributes can be reloaded, since they're not saved to the directory. The attributes `training_loop`, `training`, `model` and `random_seed` have been made optional, with `None` default values. The `stopper` (if present) is not reloaded because of some missing information (it was already optional). Other attributes such as `configuration`, `version` and `git_hash` are not reloaded, but their default value has been kept.

### Alternatives

I considered subclassing `PipelineResult` with a new class `ReloadedResult` which contained default values for the parameters that could've not been restored. However, I think this solution adds too much complexity.

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

Users should be warned about the missing information. Making the `random_seed`, `training`, `model` and `training_loop` as optional should not create any problem.
### Verification Process

<!--

What process did you follow to verify that the change has not introduced any regressions? Describe the actions you
performed and describe the results you observed.

-->

Performed tests with `tox -e py` and `tox -e docs-test` without any error. 
Tried to save and reload the results obtained via:
```
pipeline_result = pipeline(
    dataset='Nations',
    model='TransE'
)
pipeline_result.save_to_directory('nations_transe')
```
with different configurations for the boolean flags, with success. Still have to write some unit tests for the new function.

nicolafan committed:
Implement results reloading

Implement the class method `load_from_directory` to
reload a `PipelineResult` object
starting from the results saved using `save_to_directory`.

nicolafan committed:
Add documentation

nicolafan committed:
Merge branch 'pykeen:master' into load-pipeline-results

nicolafan committed:
Make model and training optional

nicolafan committed:
Merge branch 'load-pipeline-results' of https://github.com/nicolafan/pykeen into load-pipeline-results

mberr commented on src/pykeen/pipeline/api.py:
```diff
@@ -451,6 +451,62 @@ def save_to_directory(
+        
+        # load results (mandatory)
+        result_file_path = directory.joinpath(cls.RESULT_FILE_NAME)
+        if not result_file_path.exists():
```
```suggestion
        if not result_file_path.is_file():
```

[`pathlib.Path.exists`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.exists) is also True, if the path points to an existing directory.

mberr commented on src/pykeen/pipeline/api.py:
```diff
@@ -451,6 +451,62 @@ def save_to_directory(
+
+        # load metadata
+        metadata_file_path = directory.joinpath(cls.METADATA_FILE_NAME)
+        if metadata_file_path.exists():
```
see above

mberr commented on src/pykeen/pipeline/api.py:
```diff
@@ -451,6 +451,62 @@ def save_to_directory(
+
+        # load model
+        model_file_path = directory.joinpath(cls.MODEL_FILE_NAME)
+        if model_file_path.exists():
```
see above

mberr commented on src/pykeen/pipeline/api.py:
```diff
@@ -451,6 +451,62 @@ def save_to_directory(
+
+        # load training triples
+        training_triples_file_path = directory.joinpath(cls.TRAINING_TRIPLES_FILE_NAME)
+        if training_triples_file_path.exists():
```
iirc, this should be a path to a directory, so

```suggestion
        if training_triples_file_path.is_dir():
```

mberr reviewed (commented):
First of all, thanks for the PR. It looks already promising!

I quickly went over the changes and left a few comments.

nicolafan committed:
Update src/pykeen/pipeline/api.py

Convert exists check to is_file

Co-authored-by: Max Berrendorf <max.berrendorf@gmail.com>

nicolafan committed:
Update src/pykeen/pipeline/api.py

Convert exists check to is_dir

Co-authored-by: Max Berrendorf <max.berrendorf@gmail.com>

nicolafan committed:
Convert exists checks to is_file

nicolafan committed:
Save random_seed, version and git_hash

Add the possibility to save the pipeline_info, a set of
attributes of PipelineResult that were not saved in the previous
implementation of `save_to_directory`.

Reload the pipeline_info.

nicolafan committed:
Blackify

nicolafan committed:
Modify documentation

cthoyt commented on src/pykeen/pipeline/api.py:
```diff
@@ -292,6 +280,18 @@ class PipelineResult(Result):
     #: How long in seconds did evaluation take?
     evaluate_seconds: float
 
+    #: The model trained by the pipeline
```
Sorry I haven't been following discussion yet, but I'm pretty opposed to weakening the typing in this class. When people get results from the pipeline, they shouldn't have to check for if there are results there - it should be guaranteed.

This is definitely true for the model and seed. If the training triples don't get loaded properly, people get confused since they need to reference the training triples to do anything valuable. I don't want to make changes that will result in getting more github issues from confused users. Wrt the training loop, this could be a reasonable thing to make optional.

nicolafan replied:
Hi @cthoyt ! Yes, it's perfectly understandable. So, maybe we should give up the implmentation of this feature or rethink it completely if users need to reload the `PipelineResult`, since the current implementation of `save_to_directory` saves only partial information of the object.
